kernel/src/apic.c

```diff
@@ -0,0 +1,259 @@
+        KERNEL_ASSERT(madt_valid, "INVALID MADT");
+    }
+
+    // Fill Override table to have a corrent redirection when GSI = 8259 IRQ.
+    for (int i = 0; i < MAX_8259_IRQ_COUNT; i++) g_gsi_override_table[i].gsi = i;
+
+    PhysicalAddress local_apic_phys_addr = (PhysicalAddress)madt->local_apic_phys_addr;
+    { // Parse MADT
+
+        const void* end = (void*)madt + madt->header.length;
+        void* curr_addr = (void*)madt + sizeof(MADT);
+
+        // Entry offsets are found in
+        while (curr_addr < end) {
+            MADTEntry* entry_header = (MADTEntry*)curr_addr;
+            curr_addr += entry_header->length;
+
+            switch (entry_header->type) {
+                case LOCAL_APIC_ENTRY:
+                    g_found_lapics[g_lapic_count++] = (LocalAPICEntry*)entry_header;
+                    break;
+
+                case IO_APIC_ENTRY: {
+                    const IOAPICEntry* entry = (const IOAPICEntry*)entry_header;
+
+                    IOAPICInfo* ioapic = &g_found_ioapics[g_ioapic_count++];
+                    ioapic->gsi_offset = entry->global_system_interrupt_base;
+                    ioapic->ioapic_id = entry->ioapic_id;
+                    ioapic->physical_address = (PhysicalAddress)entry->ioapic_address;
+
+                    // Page each IOAPIC
+                    ioapic->virtual_address = (void*)map_range(
+                        ioapic->physical_address, 1, PAGING_WRITABLE | PAGING_CACHE_DISABLE);
+
+                    ioapic->used_irqs = 1 + read_ioapic_mre(ioapic->virtual_address);
+
+                    { // ID mismatch could indicate that the IOAPIC can't be read
+                        const uint8_t fetched_id = (uint8_t)read_ioapic_id(ioapic->virtual_address);
+                        const uint8_t stored_id = ioapic->ioapic_id;
+
+                        KERNEL_ASSERT(stored_id == fetched_id, "ID MISMATCH WHEN LOADING IOAPIC");
+                    }
+
+                    break;
+                }
+
+                case INTERRUPT_SOURCE_OVERRIDE_ENTRY: {
+                    const InterruptSourceOverrideEntry* entry =
+                        (const InterruptSourceOverrideEntry*)entry_header;
+
+                    GSIOverrideEntry* gsi_override_entry = &g_gsi_override_table[entry->irq_source];
+                    gsi_override_entry->present = true;
+                    gsi_override_entry->gsi = entry->global_system_interrupt;
+                    gsi_override_entry->flags = entry->flags;
+
+                    break;
+                }
+
+                case LOCAL_APIC_ADDRESS_OVERRIDE_ENTRY: {
+                    const LocalAPICAddressOverrideEntry* entry =
+                        (const LocalAPICAddressOverrideEntry*)entry_header;
+
+                    local_apic_phys_addr = entry->local_apic_phys_addr;
+                    break;
+                }
+
+                default: break;
+            }
+        }
+    }
+
+    { // Enable Local APIC
```

Comment:
Don't you need to set the the enable bit in the MSR to enable the local APIC?

Reply:
MSR bit is be set at boot time and cannot be disabled afterwards

Reply:
Could you link the place where you read that local APIC is always enabled at boot time? Because this osdev article seems to mention hardware enabling the local APIC: https://wiki.osdev.org/APIC#Local_APIC_configuration